libraries/node-core-library/src/EnvironmentMap.ts

```diff
@@ -0,0 +1,136 @@
+import * as process from 'process';
+import { InternalError } from './InternalError';
+
+/**
+ * A process environment variable name and its value.  Used by {@link EnvironmentMap}.
+ * @public
+ */
+export interface IEnvironmentEntry {
+  /**
+   * The name of the environment variable.
+   */
+  name: string;
+
+  /**
+   * The value of the environment variable.
+   */
+  value: string;
+}
+
+/**
+ * A map data structure that stores process environment variables.  On Windows
+ * operating system, the variable names are case-insensitive.
+ * @public
+ */
+export class EnvironmentMap {
+  private readonly _map: Map<string, IEnvironmentEntry> = new Map();
+
+  /**
+   * Whether the environment variable names are case-sensitive.
+   *
+   * @remarks
+   * On Windows operating system, environment variables are case-insensitive.
+   * The map will preserve the variable name casing from the most recent assignment operation.
+   */
+  public readonly caseSensitive: boolean;
+
+  public constructor(environmentObject: Record<string, string | undefined> = {}) {
+    // This property helps catch a mistake where an instance of `EnvironmentMap` is accidentally passed to
+    // a function that expects a `Record<string, string>` (as would be used with the `process.env` API).
+    // The property getter will throw an exception if that function tries to enumerate the object values.
+    Object.defineProperty(this, '_sanityCheck', {
+      enumerable: true,
+      get: function () {
+        throw new InternalError('Attempt to read EnvironmentMap class as an object');
+      }
+    });
+
+    this.caseSensitive = process.platform === 'win32';
+    this.mergeFromObject(environmentObject);
+  }
+
+  /**
+   * Clears all entries, resulting in an empty map.
+   */
+  public clear(): void {
+    this._map.clear();
+  }
+
+  /**
+   * Assigns the variable to the specified value.  A previous value will be overwritten.
+   *
+   * @remarks
+   * The value can be an empty string.  To completely remove the entry, use
+   * {@link EnvironmentMap.unset} instead.
+   */
+  public set(name: string, value: string): void {
+    const key: string = this.caseSensitive ? name.toUpperCase() : name;
```

Comment:
Shouldn't this be the other way around?

Reply:
😮 This is why we do code reviews heheh.  And why we write unit tests.  I'll make a test.

Reply:
😋 The unit test passed because `this.caseSensitive` is also initialized backwards.

See, this is why we are lazy about unit tests but not code reviews!